benchmarks/exp_hash.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2025 The Sigstore Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Script for running a benchmark to pick a hashing algorithm."""
+
+import argparse
+import timeit
+from typing import Final
+
+import numpy as np
+
+from model_signing.hashing import hashing
+from model_signing.hashing import memory
+
+
+KB: Final[int] = 1024
+MB: Final[int] = 1024 * KB
+GB: Final[int] = 1024 * MB
+
+
+def build_parser() -> argparse.ArgumentParser:
+    """Builds the command line parser for the hash experiment."""
+    parser = argparse.ArgumentParser(
+        description="hash algorithm benchmark data for model signing"
+    )
+
+    parser.add_argument(
+        "--repeat",
+        help="how many times to repeat each algorithm",
+        type=int,
+        default=5,
+    )
+
+    parser.add_argument(
+        "--methods",
+        help="hash methods to benchmark",
+        nargs="+",
+        type=str,
+        default=["sha256", "blake2"],
+    )
+
+    parser.add_argument(
+        "--data-sizes",
+        help="hash methods to benchmark",
+        nargs="+",
+        type=int,
+        default=[KB, MB, 512 * MB, GB, 4 * GB, 16 * GB, 32 * GB],
```

Comment:
These values are arbitrary for now. Happy to change them now/later

Reply:
[discussion] Let's use 1,2,5,10,20,50,100,200,500,... instead just of powers of 2, to get values in between. This way, an eventual plot would be more easier to understand (an issue in the old plots)


On log scale,distance between 1 and 2 is the same as between 5 and 10 and just slightly lower than between 2 and 5:

```
log10(10) - log10(5) = 0.301
log10(5) - log10(2) = 0.397
log10(2) - log10(1) = 0.301
```

These are the numbers used in [`autorange`](https://docs.python.org/3/library/timeit.html#timeit.Timer.repeat) too.

These are no longer simple powers of 2 so human readable output would change, but for plotting these sizes would be better.

But if we also want human readable sizes / powers of 2 we need different thresholds here. Let me think more about this

Reply:
Let's have a default range for human output (could be the one listed here) and another one for plots. Can make the default be empty list (I think) to signal that the user has not provided an override and then initialize it after we check for the output format flag.

Reply:
I just included the plot output for now. We can add a `-h` flag later.

`add_argument` will use a default of `None` if not provided, so I just built the logic around that.